### PR TITLE
ServiceCollection Registration updates

### PR DESCRIPTION
**Pull request by v-mserdar**
## Description
Added a new collection that has the same messages that the Debug.Assert functions were raising when services were not correctly registered. The point of this is to merely expose these messages that already provided guidance on how they should be used.
Revised the services.Add() to be services.TryAdd() to prevent duplicates getting added.
Added a unit test.

## Related issues
Addresses [92947](https://microsofthealth.visualstudio.com/Health/_workitems/edit/92947)

## Testing
Added a new unit test and locally successfully ran all tests in Microsoft.Health.Extensions.DependencyInjection.UnitTests (net6.0)

## [Semver Change](https://github.com/microsoft/healthcare-shared-components/blob/main/docs/Versioning.md)
Patch|Skip|Feature

### CODE REVIEW

**Review by wsugarman — CHANGES_REQUESTED** [review]

File: src/Microsoft.Health.Extensions.DependencyInjection/ServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -51,6 +57,25 @@ public static void RegisterAssemblyModules(this IServiceCollection collection, A
         }
     }
 
+    public static ICollection<string> CheckIssues(this IServiceCollection collection)
+    {
+        var issues = new List<string>();
+        var multipleRegisteredServices = collection
+            .GroupBy(x => (x.ServiceType, x.ImplementationType, x.ImplementationInstance, x.ImplementationFactory))
```
[the remark this answers is not in the document]
Comment (v-mserdar) [reply src/Microsoft.Health.Extensions.DependencyInjection/ServiceCollectionExtensions.cs:64]: Thanks for the feedback here. Based on the comments, especially the 3rd bullet, I should probably remove the code altogether. What started all of this was a local debug session of fhir-server hitting Debug.Asserts in TypeRegistrationBuilder.AsImplementedInterfaces() . The debug message pointed to missing AsSelf() in a handlful of areas and that's where I noticed some other debug code trying to identify duplicate service registration (which is what I added above). 
Long story short, I should probably revert/close this PR.
If you have a bug/workitem # for the Newtonsoft.json.dll CG issue, I'll use that for a new PR otherwise I'll create an item for it.

File: src/Microsoft.Health.Extensions.DependencyInjection/TypeRegistrationBuilder.cs [hunk 2]
```diff
@@ -122,9 +135,14 @@ public TypeRegistrationBuilder ReplaceService<T>()
     /// <exception cref="NotSupportedException">Throws when Type was not explicitly defined</exception>
     public TypeRegistrationBuilder AsImplementedInterfaces(Predicate<Type> interfaceFilter = null)
     {
-        Debug.Assert(
-            _firstRegisteredType != null || _registrationMode == TypeRegistration.RegistrationMode.Transient,
-            $"Using \"AsImplementedInterfaces()\" without calling \"AsSelf()\" first in registration for \"{_type.Name}\" could have inconsistent results.");
+        bool assertion = _firstRegisteredType != null || _registrationMode == TypeRegistration.RegistrationMode.Transient;
+        string msg = $"Using \"AsImplementedInterfaces()\" without calling \"AsSelf()\" first in registration for \"{_type.Name}\" could have inconsistent results.";
```
Comment (wsugarman) [inline src/Microsoft.Health.Extensions.DependencyInjection/TypeRegistrationBuilder.cs:139]: Instead of warnings, should there be a setting on the `TypeRegistrationBuilder` when first created that enforces these assertions, like:
```csharp
IServiceCollection services = new ServiceCollection();
services
    .Add<T>(strictRegistration: true) // Or whatever name we like lol
    .AsSelf()
    .AsInterfaces();
```

Alternatively, this sort of detection would probably be a good candidate for a [custom Roslyn Analyzer](https://devblogs.microsoft.com/dotnet/how-to-write-a-roslyn-analyzer/)

**Review by wsugarman — CHANGES_REQUESTED** [review]

File: src/Microsoft.Health.Extensions.DependencyInjection/ServiceCollectionExtensions.cs [hunk 3]
```diff
@@ -29,8 +29,14 @@ public static void RegisterModule<T>(this IServiceCollection collection, params
         EnsureArg.IsNotNull(collection, nameof(collection));
 
         var moduleType = typeof(T);
-        Debug.Assert(moduleType.IsClass && !moduleType.IsAbstract, "Module should be non-abstract class");
-        collection.RegisterModule(moduleType, constructorParams);
+        bool assertion = moduleType.IsClass && !moduleType.IsAbstract;
+        Debug.Assert(assertion, "Module should be non-abstract class");
+
+        // have the registration follow same rules as fn RegisterAssemblyModules
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Extensions.DependencyInjection/ServiceCollectionExtensions.cs:35]: Nit: Typo, fn

File: src/Microsoft.Health.Extensions.DependencyInjection/TypeRegistrationBuilder.cs (hunk 2 above)
Comment (LTA-Thinking) [reply src/Microsoft.Health.Extensions.DependencyInjection/TypeRegistrationBuilder.cs:139]: Instead of a Debug.Assert could we have it throw an exception? That way it would be still be caught by components downstream. I agree with will a strictMode or similar setting would be useful, but by default set it to be enabled. If there is a case where we know we need the "wrong" behavior we can override the exception check.

**Comment by jovinson-ms** [comment]
Closed after confirming with @v-mserdar.
